Sources/LSPLogging/SetGlobalLogFileHandler.swift

```diff
@@ -0,0 +1,188 @@
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import RegexBuilder
+
+#if !canImport(os) || SOURCEKITLSP_FORCE_NON_DARWIN_LOGGER
+fileprivate struct FailedToCreateFileError: Error, CustomStringConvertible {
+  let logFile: URL
+
+  var description: String {
+    return "Failed to create log file at \(logFile)"
+  }
+}
+
+/// The number of log file handles that have been created by this process.
+///
+/// See comment on `logFileHandle`.
+@LogHandlerActor
+fileprivate var logRotateIndex = 0
+
+/// The file handle to the current log file. When the file managed by this handle reaches its maximum size, we increment
+/// the `logRotateIndex` by 1 and set the `logFileHandle` to `nil`. This causes a new log file handle with index
+/// `logRotateIndex % logRotateCount` to be created on the next log call.
+@LogHandlerActor
+fileprivate var logFileHandle: FileHandle?
+
+@LogHandlerActor
+func getOrCreateLogFileHandle(logDirectory: URL, logRotateCount: Int) -> FileHandle {
+  if let logFileHandle {
+    return logFileHandle
+  }
+
+  // Name must match the regex in `cleanOldLogFiles` and the prefix in `DiagnoseCommand.addNonDarwinLogs`.
+  let logFileUrl = logDirectory.appendingPathComponent(
+    "sourcekit-lsp-\(ProcessInfo.processInfo.processIdentifier).\(logRotateIndex % logRotateCount).log"
+  )
+
+  do {
+    try FileManager.default.createDirectory(at: logDirectory, withIntermediateDirectories: true)
+    if !FileManager.default.fileExists(atPath: logFileUrl.path) {
+      guard FileManager.default.createFile(atPath: logFileUrl.path, contents: nil) else {
+        throw FailedToCreateFileError(logFile: logFileUrl)
+      }
+    }
+    let newFileHandle = try FileHandle(
+      forWritingTo: logDirectory.appendingPathComponent(
+        "sourcekit-lsp-\(ProcessInfo.processInfo.processIdentifier).\(logRotateIndex % logRotateCount).log"
+      )
+    )
+    logFileHandle = newFileHandle
+    try newFileHandle.truncate(atOffset: 0)
+    return newFileHandle
+  } catch {
+    // If we fail to create a file handle for the log file, log one message about it to stderr and then log to stderr.
+    // We will try creating a log file again once this section of the log reaches `maxLogFileSize` but that means that
+    // we'll only log this error every `maxLogFileSize` bytes, which is a lot less spammy than logging it on every log
+    // call.
+    fputs("Failed to open file handle for log file at \(logFileUrl.path): \(error)", stderr)
+    logFileHandle = FileHandle.standardError
+    return FileHandle.standardError
+  }
+}
+
+/// Log the given message to the given log file handle.
+///
+/// Occasionally check if the log file exceeds the target log size. If this is the case, truncate the beginning of the
+/// log to reduce the log size.
```

Comment:
I guess you could also say that the logs will be rotated if `logRotateCount > 1`

Reply:
Oh, this comment is just completely outdated.